homeassistant/components/upcloud/__init__.py

```diff
@@ -235,3 +248,13 @@ def extra_state_attributes(self) -> dict[str, Any]:
                 ATTR_MEMORY_AMOUNT,
             )
         }
+
+    @property
+    def device_info(self) -> DeviceInfo:
+        """Return info for device registry."""
+        assert self.coordinator.config_entry is not None
```

Comment:
An `assert` statement shouldn't be used in production code as it will be removed when compilation is optimized. If it's not possible to ensure that `self.coordinator.config_entry` is always not `None`you should raise an exception instead.

Reply:
It should just return None in case the entry is unknown.

Reply:
Using `assert` is one valid way we use a lot to make type checks pass, arguably better than a cast or a type ignore. Anyway in this case returning `None` seems to be an option indeed, will have a look.

Reply:
It appears other implementations that use config entry data in the device_info they return contain a custom data update coordinator to which a non-None config entry is passed on init. This seems wasteful to me, as there already is a config entry in the base class, it's just type hinted so that it can be None (in `config_entries.current_entry`) -- and for cases like the one at hand, I doubt it can actually be None because the related config entry surely was initialized and thus current_entry was set.

So in that sense the assert seems fine to me, no need to litter the code with other more verbose and misleading workarounds for things that don't actually happen in practice.